<!DOCTYPE html>
<html>
  <head>
    <title>Data manipulation with dplyr, purrr and broom</title>
    <meta charset="utf-8">
    <meta name="author" content="Di Cook (dicook@monash.edu, @visnut)   Heike Hofmann (heike.hofmann@gmail.com, @heike_hh)" />
    <link href="index_files/remark-css-0.0.1/example.css" rel="stylesheet" />
    <link rel="stylesheet" href="myremark.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Data manipulation with dplyr, purrr and broom
## SISBID 2017 <br> <a href="https://github.com/SISBID/Module2" class="uri">https://github.com/SISBID/Module2</a>
### Di Cook (<a href="mailto:dicook@monash.edu">dicook@monash.edu</a>, <span class="citation">@visnut</span>) <br> Heike Hofmann (<a href="mailto:heike.hofmann@gmail.com">heike.hofmann@gmail.com</a>, <span class="citation">@heike_hh</span>)
### 07/12-14/2017

---







# Outline

- dplyr package: motivation, functions, chaining
- purrr: working with lists, vectors of data frames

---
# dplyr verbs

There are five primary dplyr **verbs**, representing distinct data analysis tasks:

- Filter: Remove the rows of a data frame, producing subsets
- Arrange: Reorder the rows of a data frame
- Select: Select particular columns of a data frame
- Mutate: Add new columns that are functions of existing columns
- Summarise: Create collapsed summaries of a data frame
- (Group By: Introduce structure to a data frame)
 
 
---
# Filter



```r
data(french_fries, package = "reshape2")
french_fries %&gt;%
    filter(subject == 3, time == 1)
  time treatment subject rep potato buttery grassy rancid painty
1    1         1       3   1    2.9     0.0    0.0    0.0    5.5
2    1         1       3   2   14.0     0.0    0.0    1.1    0.0
3    1         2       3   1   13.9     0.0    0.0    3.9    0.0
4    1         2       3   2   13.4     0.1    0.0    1.5    0.0
5    1         3       3   1   14.1     0.0    0.0    1.1    0.0
6    1         3       3   2    9.5     0.0    0.6    2.8    0.0
```

---
# Arrange


```r
french_fries %&gt;%
    arrange(desc(rancid)) %&gt;%
    head
  time treatment subject rep potato buttery grassy rancid painty
1    9         2      51   1    7.3     2.3      0   14.9    0.1
2   10         1      86   2    0.7     0.0      0   14.3   13.1
3    5         2      63   1    4.4     0.0      0   13.8    0.6
4    9         2      63   1    1.8     0.0      0   13.7   12.3
5    5         2      19   2    5.5     4.7      0   13.4    4.6
6    4         3      63   1    5.6     0.0      0   13.3    4.4
```

---
# Select


```r
french_fries %&gt;%
    select(time, treatment, subject, rep, potato) %&gt;%
    head
   time treatment subject rep potato
61    1         1       3   1    2.9
25    1         1       3   2   14.0
62    1         1      10   1   11.0
26    1         1      10   2    9.9
63    1         1      15   1    1.2
27    1         1      15   2    8.8
```

---
# Summarise


```r
french_fries %&gt;%
    summarise(mean_rancid = mean(rancid, na.rm=TRUE), 
              sd_rancid = sd(rancid, na.rm = TRUE))
  mean_rancid sd_rancid
1     3.85223  3.781815
```

---
# Summarise and Group_by


```r
french_fries %&gt;%
    group_by(time, treatment) %&gt;%
    summarise(mean_rancid = mean(rancid), sd_rancid = sd(rancid))
# A tibble: 30 x 4
# Groups:   time [?]
     time treatment mean_rancid sd_rancid
   &lt;fctr&gt;    &lt;fctr&gt;       &lt;dbl&gt;     &lt;dbl&gt;
 1      1         1    2.758333  3.212870
 2      1         2    1.716667  2.714801
 3      1         3    2.600000  3.202037
 4      2         1    3.900000  4.374730
 5      2         2    2.141667  3.117540
 6      2         3    2.495833  3.378767
 7      3         1    4.650000  3.933358
 8      3         2    2.895833  3.773532
 9      3         3    3.600000  3.592867
10      4         1    2.079167  2.394737
# ... with 20 more rows
```



---
# Let's use these tools

to answer these french fry experiment questions:

- Is the design complete?
- Are replicates like each other?
- How do the ratings on the different scales differ?
- Are raters giving different scores on average?
- Do ratings change over the weeks?

---
# Completeness 
If the data is complete it should be 12 x 10 x 3 x 2, that is, 6 records for each person. (Assuming that each person rated on all scales.) 

To check this we want to tabulate the number of records for each subject, time and treatment. This means select appropriate columns, tabulate, count and spread it out to give a nice table.

---
# 


```r
french_fries %&gt;% 
#  select(subject, time, treatment) %&gt;% 
  count(subject, time) %&gt;%
  spread(time, n)
# A tibble: 12 x 11
   subject   `1`   `2`   `3`   `4`   `5`   `6`   `7`   `8`   `9`  `10`
 *  &lt;fctr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
 1       3     6     6     6     6     6     6     6     6     6    NA
 2      10     6     6     6     6     6     6     6     6     6     6
 3      15     6     6     6     6     6     6     6     6     6     6
 4      16     6     6     6     6     6     6     6     6     6     6
 5      19     6     6     6     6     6     6     6     6     6     6
 6      31     6     6     6     6     6     6     6     6    NA     6
 7      51     6     6     6     6     6     6     6     6     6     6
 8      52     6     6     6     6     6     6     6     6     6     6
 9      63     6     6     6     6     6     6     6     6     6     6
10      78     6     6     6     6     6     6     6     6     6     6
11      79     6     6     6     6     6     6     6     6     6    NA
12      86     6     6     6     6     6     6     6     6    NA     6
```

---
# Check completeness with different scales, too


```r
french_fries %&gt;% 
  gather(type, rating, -subject, -time, -treatment, -rep) %&gt;%
  select(subject, time, treatment, type) %&gt;% 
  count(subject, time) %&gt;%
  spread(time, n)
# A tibble: 12 x 11
   subject   `1`   `2`   `3`   `4`   `5`   `6`   `7`   `8`   `9`  `10`
 *  &lt;fctr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
 1       3    30    30    30    30    30    30    30    30    30    NA
 2      10    30    30    30    30    30    30    30    30    30    30
 3      15    30    30    30    30    30    30    30    30    30    30
 4      16    30    30    30    30    30    30    30    30    30    30
 5      19    30    30    30    30    30    30    30    30    30    30
 6      31    30    30    30    30    30    30    30    30    NA    30
 7      51    30    30    30    30    30    30    30    30    30    30
 8      52    30    30    30    30    30    30    30    30    30    30
 9      63    30    30    30    30    30    30    30    30    30    30
10      78    30    30    30    30    30    30    30    30    30    30
11      79    30    30    30    30    30    30    30    30    30    NA
12      86    30    30    30    30    30    30    30    30    NA    30
```

---
# Change in ratings over weeks, relative to experimental design


```r
ff.m &lt;- french_fries %&gt;% 
  gather(type, rating, -subject, -time, -treatment, -rep)
ggplot(data=ff.m, aes(x=time, y=rating, colour=treatment)) +
  geom_point() +
  facet_grid(subject~type) 
```

---


&lt;img src="index_files/figure-html/unnamed-chunk-11-1.png" style="display: block; margin: auto;" /&gt;

---
# Add means over reps, and connect the dots


```r
ff.m.av &lt;- ff.m %&gt;% 
  group_by(subject, time, type, treatment) %&gt;%
  summarise(rating=mean(rating))
ggplot(data=ff.m, aes(x=time, y=rating, colour=treatment)) + 
  facet_grid(subject~type) +
  geom_line(data=ff.m.av, aes(group=treatment))
```

---

&lt;img src="index_files/figure-html/unnamed-chunk-13-1.png" style="display: block; margin: auto;" /&gt;

---
class: inverse middle 
# Your turn

![](lorikeets.png)

Write an answer to each of the questions:

- Is the design complete?
- Are replicates like each other?
- How do the ratings on the different scales differ?
- Are raters giving different scores on average?
- Do ratings change over the weeks?

---
# Working with lots of models

Why would we even do that?

- Hans Rosling can explain that really well in his [TED talk](https://www.ted.com/talks/hans_rosling_shows_the_best_stats_you_ve_ever_seen?language=en)
- [gapminder project](https://www.gapminder.org/)
- gapminder also makes data available (in R through the package `gapminder`)

---
# First Look


```r
library(gapminder)
ggplot(data=gapminder, aes(x=year, y=lifeExp, group=country)) + 
  geom_line()
```

&lt;img src="index_files/figure-html/unnamed-chunk-14-1.png" style="display: block; margin: auto;" /&gt;

How would you describe this plot?

---
# First Look


```r
library(gapminder)
ggplot(data=gapminder, aes(x=year, y=lifeExp, group=country, colour=continent)) + 
  geom_line()
```

&lt;img src="index_files/figure-html/unnamed-chunk-15-1.png" style="display: block; margin: auto;" /&gt;

How about now?

---
# Using models as exploratory tools

- Idea: fit a line to each one of the countries life expectancies
- then use e.g. intercept and slope to characterise groups of countries

&lt;img src="index_files/figure-html/unnamed-chunk-16-1.png" style="display: block; margin: auto;" /&gt;


---
# Life Expectancy in the U.S.

Start with a single country, then generalize over all countries: 


```r
usa &lt;- gapminder %&gt;% filter(country=="United States")
head(usa)
# A tibble: 6 x 6
        country continent  year lifeExp       pop gdpPercap
         &lt;fctr&gt;    &lt;fctr&gt; &lt;int&gt;   &lt;dbl&gt;     &lt;int&gt;     &lt;dbl&gt;
1 United States  Americas  1952   68.44 157553000  13990.48
2 United States  Americas  1957   69.49 171984000  14847.13
3 United States  Americas  1962   70.21 186538000  16173.15
4 United States  Americas  1967   70.76 198712000  19530.37
5 United States  Americas  1972   71.34 209896000  21806.04
6 United States  Americas  1977   73.38 220239000  24072.63
```

---
# Life Expectancy in the U.S. since 1950


```r
ggplot(data=usa, aes(x=year, y=lifeExp)) + 
  geom_point() +
  geom_smooth(method="lm")
```

&lt;img src="index_files/figure-html/unnamed-chunk-18-1.png" style="display: block; margin: auto;" /&gt;

---
# Life Expectancy in the U.S. since 1950


```r
usa.lm &lt;- lm(lifeExp~year, data=usa)
usa.lm

Call:
lm(formula = lifeExp ~ year, data = usa)

Coefficients:
(Intercept)         year  
  -291.0845       0.1842  
```

Intercept is estimated life expectancy at 0 BC - let's use 1950 for the first value:


```r
gapminder &lt;- gapminder %&gt;% mutate(year = year-1950)
```

---
# Nesting data

We don't want to subset the data for every country ... 

`nest()` makes a data frame part of another data frame:


```r
by_country &lt;- gapminder %&gt;% group_by(continent, country) %&gt;% nest()
head(by_country)
# A tibble: 6 x 3
  continent     country              data
     &lt;fctr&gt;      &lt;fctr&gt;            &lt;list&gt;
1      Asia Afghanistan &lt;tibble [12 x 4]&gt;
2    Europe     Albania &lt;tibble [12 x 4]&gt;
3    Africa     Algeria &lt;tibble [12 x 4]&gt;
4    Africa      Angola &lt;tibble [12 x 4]&gt;
5  Americas   Argentina &lt;tibble [12 x 4]&gt;
6   Oceania   Australia &lt;tibble [12 x 4]&gt;
```

Each element of the `data` variable in `by_country` is a dataset.

We will fit a linear model to each one of the elements of the `by_country` data.

---

--- 
# Fitting multiple models

Now we are using the `map` function in the package `purrr`.

`map` allows us to apply a function to each element of a list.


```r
by_country$model &lt;- by_country$data %&gt;% purrr::map(~lm(lifeExp~year, data=.))
head(by_country)
# A tibble: 6 x 4
  continent     country              data    model
     &lt;fctr&gt;      &lt;fctr&gt;            &lt;list&gt;   &lt;list&gt;
1      Asia Afghanistan &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt;
2    Europe     Albania &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt;
3    Africa     Algeria &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt;
4    Africa      Angola &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt;
5  Americas   Argentina &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt;
6   Oceania   Australia &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt;
```

---
# `broom` again

using `broom::tidy` we extract coefficients for each model


```r
broom::tidy(by_country$model[[1]])
         term   estimate  std.error statistic      p.value
1 (Intercept) 29.3566375 0.69898128  41.99918 1.404235e-12
2        year  0.2753287 0.02045093  13.46289 9.835213e-08
```

---

--- 
# Extract values for each coefficient

Extract all countries automatically (hello `map` again!)


```r
by_country$coefs &lt;- by_country$model %&gt;% purrr::map(broom::tidy)
by_country
# A tibble: 142 x 5
   continent     country              data    model                coefs
      &lt;fctr&gt;      &lt;fctr&gt;            &lt;list&gt;   &lt;list&gt;               &lt;list&gt;
 1      Asia Afghanistan &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt; &lt;data.frame [2 x 5]&gt;
 2    Europe     Albania &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt; &lt;data.frame [2 x 5]&gt;
 3    Africa     Algeria &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt; &lt;data.frame [2 x 5]&gt;
 4    Africa      Angola &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt; &lt;data.frame [2 x 5]&gt;
 5  Americas   Argentina &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt; &lt;data.frame [2 x 5]&gt;
 6   Oceania   Australia &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt; &lt;data.frame [2 x 5]&gt;
 7    Europe     Austria &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt; &lt;data.frame [2 x 5]&gt;
 8      Asia     Bahrain &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt; &lt;data.frame [2 x 5]&gt;
 9      Asia  Bangladesh &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt; &lt;data.frame [2 x 5]&gt;
10    Europe     Belgium &lt;tibble [12 x 4]&gt; &lt;S3: lm&gt; &lt;data.frame [2 x 5]&gt;
# ... with 132 more rows
```

---

--- 
# Get the coefficients into one data set

`unnest` pulls little data frames back into the whole frame


```r
by_country_coefs &lt;- by_country %&gt;% unnest(coefs)
by_country_coefs
# A tibble: 284 x 7
   continent     country        term   estimate   std.error  statistic
      &lt;fctr&gt;      &lt;fctr&gt;       &lt;chr&gt;      &lt;dbl&gt;       &lt;dbl&gt;      &lt;dbl&gt;
 1      Asia Afghanistan (Intercept) 29.3566375 0.698981278  41.999176
 2      Asia Afghanistan        year  0.2753287 0.020450934  13.462890
 3    Europe     Albania (Intercept) 58.5597618 1.133575812  51.659325
 4    Europe     Albania        year  0.3346832 0.033166387  10.091036
 5    Africa     Algeria (Intercept) 42.2364149 0.756269040  55.848399
 6    Africa     Algeria        year  0.5692797 0.022127070  25.727749
 7    Africa      Angola (Intercept) 31.7079741 0.804287463  39.423683
 8    Africa      Angola        year  0.2093399 0.023532003   8.895964
 9  Americas   Argentina (Intercept) 62.2250191 0.167091314 372.401279
10  Americas   Argentina        year  0.2317084 0.004888791  47.395847
# ... with 274 more rows, and 1 more variables: p.value &lt;dbl&gt;
```

---

--- 
# ... and finally, a visualization:


```r
coefs &lt;- by_country_coefs %&gt;% 
  select(country, continent, term, estimate) %&gt;% 
  spread(term, estimate)
ggplot(data=coefs, aes(x=`(Intercept)`, y=year, colour=continent)) +
  geom_point()
```

&lt;img src="index_files/figure-html/unnamed-chunk-26-1.png" style="display: block; margin: auto;" /&gt;

---
class: inverse middle 
# Your turn

![](lorikeets.png)

- use `broom::glance` to get model fits for each one of the linear models of life expectancy by year
- which countries have the best five fits? which countries have the worst fits?
- are there meaningful patterns?




---
class: inverse middle 
# Your turn

![](lorikeets.png)

- extract residuals for each of the models and store it in a dataset together with country and continent information
- plot residuals across the years and fit a smooth. What does the pattern mean?



---
# broom for bio data

`biobroom` package (Bioconductor; Bass, Nelson, Robinson, Storey, 2016) has the same functions as `broom`, i.e. `glance`, `tidy`, and `augment`.

These functions provide information depending on the input type


```r
library(biobroom)
data(hammer)

counts &lt;- Biobase::exprs(hammer)
head(counts)
                   SRX020102 SRX020103 SRX020104 SRX020105 SRX020091-3
ENSRNOG00000000001         2         4        18        24           7
ENSRNOG00000000007         4         1         3         1           5
ENSRNOG00000000008         0         1         4         2           0
ENSRNOG00000000009         0         0         0         0           0
ENSRNOG00000000010        19        10        19        13          50
ENSRNOG00000000012         7         5         1         0          31
                   SRX020088-90 SRX020094-7 SRX020098-101
ENSRNOG00000000001            4          93            77
ENSRNOG00000000007            4           9             4
ENSRNOG00000000008            5           2             6
ENSRNOG00000000009            0           0             0
ENSRNOG00000000010           57          45            58
ENSRNOG00000000012           26          12             9
```

---
# The Hammer study

published as part of the `biobroom` package, part of the [ReCount project](http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-449)

```r
# more information about the study:
Biobase::phenoData(hammer)@data
                  sample.id num.tech.reps protocol         strain     Time
SRX020102         SRX020102             1  control Sprague Dawley 2 months
SRX020103         SRX020103             2  control Sprague Dawley 2 months
SRX020104         SRX020104             1   L5 SNL Sprague Dawley 2 months
SRX020105         SRX020105             2   L5 SNL Sprague Dawley  2months
SRX020091-3     SRX020091-3             1  control Sprague Dawley  2 weeks
SRX020088-90   SRX020088-90             2  control Sprague Dawley  2 weeks
SRX020094-7     SRX020094-7             1   L5 SNL Sprague Dawley  2 weeks
SRX020098-101 SRX020098-101             2   L5 SNL Sprague Dawley  2 weeks
```

---
# biobroom and edgeR

identify differentially expressed genes following the protocol by Robinson, McCarthy and Smyth 2009


```r
library(edgeR)
y &lt;- DGEList(counts = counts, group=Biobase::phenoData(hammer)@data$protocol)
y &lt;- calcNormFactors(y)
y &lt;- estimateCommonDisp(y)
y &lt;- estimateTagwiseDisp(y)
et &lt;- exactTest(y)

glance(et, alpha = 0.05) # glance on DGEExact
  significant     comparison
1        6993 control/L5 SNL
```

---
# biobroom and edgeR


```r
tet &lt;- tidy(et)
tet$significant &lt;- tet$p.value &lt; 0.05
ggplot(data=tet, aes(x=logCPM, y=estimate, colour=significant)) +
  geom_point(alpha=.1) + facet_wrap(~significant, labeller="label_both")
```

&lt;img src="index_files/figure-html/unnamed-chunk-32-1.png" style="display: block; margin: auto;" /&gt;

`augment(y)` only works for DGEList objects


---
# biobroom and limma


```r
bb &lt;- data.frame(read_tsv("../data/biotin-rma2.txt"))
head(data.frame(bb[,-2]))
        Gene biotin.WT.01.1 biotin.bio101.4 biotin.WT.B1.2 biotin.bio1B1.3
1   11986_at       6.359180        6.004075       6.325338        6.255888
2   11987_at       7.833549        6.666034       7.738628        7.691321
3   11988_at       6.145599        6.128749       6.258157        6.353912
4   11989_at       5.675039        5.078695       5.444967        5.380756
5   11990_at       5.382078        5.241411       5.209638        5.357740
6 11991_g_at       5.389809        5.087797       5.403933        5.592901
  biotin.WT.02.1 biotin.bio102.4 biotin.WT.B2.2 biotin.bio1B2.3
1       6.554727        6.144557       6.098169        6.201624
2       7.461711        7.300383       7.353033        7.298990
3       6.097703        6.183593       6.184723        6.276867
4       5.258524        5.328962       5.460898        5.203674
5       5.337013        5.276889       5.434531        5.362002
6       5.571362        5.249382       5.596683        5.140585
row.names(bb) &lt;- bb$Gene
```

---
# Looking at the gene expression data


```r
ggpairs(bb, columns=c(3,7,4,8))
```

&lt;img src="index_files/figure-html/unnamed-chunk-34-1.png" style="display: block; margin: auto;" /&gt;

---
# A porcupine plot again


```r
sub &lt;- bb %&gt;% select(Gene, biotin.WT.01.1, biotin.WT.02.1, biotin.bio101.4, biotin.bio102.4)
ggplot(sub, aes(x=biotin.WT.01.1, xend=biotin.WT.02.1, y=biotin.bio101.4, yend=biotin.bio102.4)) +
  geom_segment() +
  theme(aspect.ratio = 1) +
  xlab("wildtype, control treatment") +
  ylab("mutant, treated")
```

&lt;img src="index_files/figure-html/unnamed-chunk-35-1.png" style="display: block; margin: auto;" /&gt;

---
# Fit a limma model


```r
design &lt;- expand.grid(type=c("wild", "mutant"), trt=c("control", "treatment"), rep=1:2)

fit &lt;- lmFit(bb[,-(1:2)], model.matrix(~ type*trt, design))
fit &lt;- eBayes(fit)

head(topTable(fit))
           typemutant trttreatment typemutant.trttreatment  AveExpr
13212_s_at   3.436575    0.1501056               -3.282545 6.383540
13449_at     2.498316    0.1163889               -2.488131 5.287509
14609_at     2.301786   -0.2724221               -1.962619 5.406537
16016_at    -3.749019    1.2864088                3.338380 7.858121
18255_at     1.696560   -0.3977035               -1.326075 6.844090
15162_at     2.699042    0.1022273               -2.354713 8.408859
                   F      P.Value    adj.P.Val
13212_s_at 297.14511 8.048043e-08 0.0006677461
13449_at   154.05666 8.051126e-07 0.0033400095
14609_at   129.24605 1.484583e-06 0.0041058604
16016_at   108.19217 2.752579e-06 0.0048917219
18255_at   106.07250 2.947886e-06 0.0048917219
15162_at    91.58383 4.898100e-06 0.0067732558
```

---
class: inverse middle 
# Your Turn

![](lorikeets.png)

- For the previous example, try out what output the different broom functions (`glance`, `tidy`, `augment`)  produce.
- Create a Volcano plot for each of the model terms, i.e. plot estimates on x by log(p.values) on y. 
- Are there differences visible between the terms?



---


```r
bbfit &lt;- tidy(fit)
ggplot(data=bbfit, aes(x=term, y=estimate, group=gene)) +
  geom_line(alpha=0.1) +
  geom_point(aes(color=log(p.value)), size=2, alpha=0.6)
```

&lt;img src="index_files/figure-html/unnamed-chunk-38-1.png" style="display: block; margin: auto;" /&gt;

---

Is type*treatment interaction necessary? Very strong negative correlation is suspicious.



```r
bbfit_m &lt;- bbfit %&gt;% select(gene, term, estimate, p.value) %&gt;% 
  gather(fit.stat, value, -gene, -term) %&gt;%
  unite(term_stat, term, fit.stat) %&gt;%
  spread(term_stat, value) %&gt;% 
  rename(trt=trttreatment_estimate, mut=typemutant_estimate,
         int=`typemutant:trttreatment_estimate`, 
         trtp=trttreatment_p.value, mutp=typemutant_p.value, 
         intp=`typemutant:trttreatment_p.value`) 
ggpairs(bbfit_m, columns=c(2,4,6), upper=list(continuous="points"),
          ggplot2::aes(colour=intp)) + theme(aspect.ratio=1)
```

---

&lt;img src="index_files/figure-html/unnamed-chunk-40-1.png" style="display: block; margin: auto;" /&gt;

---


```r
fit2 &lt;- lmFit(bb[,-(1:2)], model.matrix(~ type+trt, design))
fit2 &lt;- eBayes(fit2)
bbfit2 &lt;- tidy(fit2)
ggplot(data=bbfit2, aes(x=term, y=estimate, group=gene)) +
  geom_line(alpha=0.1) +
  geom_point(aes(color=log(p.value)), size=2, alpha=0.6)
```

&lt;img src="index_files/figure-html/unnamed-chunk-41-1.png" style="display: block; margin: auto;" /&gt;

---
# References

- Ian Lyttle's notes on using [purrr ](http://ijlyttle.github.io/isugg_purrr/presentation.html#(1)
- RStudio's tutorial by Charlotte Wickham on using [purrr](https://www.rstudio.com/resources/videos/happy-r-users-purrr-tutorial/)
- [R blogger's article on purrr](http://www.r-bloggers.com/purrr-0-1-0/)
- Data transformation [cheatsheet](https://github.com/rstudio/cheatsheets/raw/master/source/pdfs/data-transformation-cheatsheet.pdf)
- [Wickham (2007) Reshaping data](https://www.jstatsoft.org/article/view/v021i12/v21i12.pdf)
- [broom vignettes, David Robinson](https://cran.r-project.org/web/packages/broom/vignettes/broom.html)
- [Wickham (2011) Split-Apply-Combine](https://www.jstatsoft.org/article/view/v040i01)

---
# Share and share alike

This work is licensed under the Creative Commons Attribution-Noncommercial 3.0 United States License. To view a copy of this license, visit http://creativecommons.org/licenses/by-nc/3.0/us/ or send a letter to Creative Commons, 171 Second Street, Suite 300, San Francisco, California, 94105, USA.
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('showSlide', function (slide) {setTimeout(function() {window.dispatchEvent(new Event('resize'));}, 100)});</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
