---
title: "Data manipulation with dplyr, purrr and broom"
subtitle: "SISBID 2017 <br> https://github.com/SISBID/Module2"
author: "Di Cook (dicook@monash.edu, @visnut) <br> Heike Hofmann (heike.hofmann@gmail.com, @heike_hh)"
date: "07/12-14/2017"
output:
  xaringan::moon_reader:
    css: ["default", "myremark.css"]
    self_contained: false
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  collapse = TRUE,
  comment = "",
  fig.height = 4,
  fig.width = 8,
  fig.align = "center",
  cache = FALSE
)
```


```{r, echo = FALSE}
library(ggplot2)
library(tidyr)
library(dplyr)
library(lubridate)
library(scales)
library(readr)
library(ggmap)
library(HLMdiag)
library(RColorBrewer)
library(gridExtra)
library(GGally)
```

# Outline

- `dplyr` package: motivation, functions, chaining
- `purrr`: working with list variables, vectors of data frames

---
# dplyr verbs

There are five primary dplyr **verbs**, representing distinct data analysis tasks:

- `filter`: Remove the rows of a data frame, producing subsets
- `arrange`: Reorder the rows of a data frame
- `select`: Select particular columns of a data frame
- `mutate`: Add new columns that are functions of existing columns
- `summarise`: Create collapsed summaries of a data frame
- (`group_by`: Introduce structure to a data frame)
 
 
---
# Filter

select a subset of the observations (horizontal selection)
```{r}
data(french_fries, package = "reshape2")
french_fries %>%
    filter(subject == 3, time == 1)
```

---
# Arrange

order the observations (hierarchically)
```{r}
french_fries %>%
    arrange(desc(rancid)) %>%
    head
```

---
# Select

select a subset of the variables (vertical selection)
```{r}
french_fries %>%
    select(time, treatment, subject, rep, potato) %>%
    head
```

---
# Summarise

summarize observations into a (set of) one-number statistic(s):
```{r}
french_fries %>%
    summarise(mean_rancid = mean(rancid, na.rm=TRUE), 
              sd_rancid = sd(rancid, na.rm = TRUE))
```

---
# Summarise and Group_by

```{r}
french_fries %>%
    group_by(time, treatment) %>%
    summarise(mean_rancid = mean(rancid), sd_rancid = sd(rancid))
```



---
# Let's use these tools

to answer these french fry experiment questions:

- Is the design complete?
- Are replicates like each other?
- How do the ratings on the different scales differ?
- Are raters giving different scores on average?
- Do ratings change over the weeks?

---
# Completeness 
If the data is complete it should be 12 x 10 x 3 x 2, that is, 6 records for each person. (Assuming that each person rated on all scales.) 

To check this we want to tabulate the number of records for each subject, time and treatment. This means select appropriate columns, tabulate, count and spread it out to give a nice table.

---
# 

```{r}
french_fries %>% 
#  select(subject, time, treatment) %>% 
  count(subject, time) %>%
  spread(time, n)
```

---
# Check completeness with different scales, too

```{r}
french_fries %>% 
  gather(type, rating, -subject, -time, -treatment, -rep) %>%
  select(subject, time, treatment, type) %>% 
  count(subject, time) %>%
  spread(time, n)
```

---
# Change in ratings over weeks, relative to experimental design

```{r fig.show='hide'}
ff.m <- french_fries %>% 
  gather(type, rating, -subject, -time, -treatment, -rep)
ggplot(data=ff.m, aes(x=time, y=rating, colour=treatment)) +
  geom_point() +
  facet_grid(subject~type) 
```

---


```{r echo=FALSE, fig.width=12, fig.height=9}
ggplot(data=ff.m, aes(x=time, y=rating, colour=treatment)) +
  geom_point() +
  facet_grid(subject~type) 
```

---
# Add means over reps, and connect the dots

```{r fig.show='hide'}
ff.m.av <- ff.m %>% 
  group_by(subject, time, type, treatment) %>%
  summarise(rating=mean(rating))
ggplot(data=ff.m, aes(x=time, y=rating, colour=treatment)) + 
  facet_grid(subject~type) +
  geom_line(data=ff.m.av, aes(group=treatment))
```

---

```{r echo=FALSE, fig.width=12, fig.height=9}
ggplot(data=ff.m, aes(x=time, y=rating, colour=treatment)) + 
  facet_grid(subject~type) +
  geom_line(data=ff.m.av, aes(group=treatment))
```

---
class: inverse middle 
# Your turn

![](lorikeets.png)

Write an answer to each of the questions:

- Is the design complete?
- Are replicates like each other?
- How do the ratings on the different scales differ?
- Are raters giving different scores on average?
- Do ratings change over the weeks?

---
# Working with lots of models

Why would we even do that?

- Hans Rosling can explain that really well in his [TED talk](https://www.ted.com/talks/hans_rosling_shows_the_best_stats_you_ve_ever_seen?language=en)
- [gapminder project](https://www.gapminder.org/)
- gapminder also makes data available (in R through the package `gapminder`)

---
# First Look

```{r}
library(gapminder)
ggplot(data=gapminder, 
       aes(x=year, y=lifeExp, group=country)) + 
  geom_line()
```

How would you describe this plot?

---
# First Look

```{r}
library(gapminder)
ggplot(data=gapminder, 
       aes(x=year, y=lifeExp, group=country, colour=continent)) + 
  geom_line()
```

How about now?

---
# Using models as exploratory tools

- Idea: fit a line to each one of the countries life expectancies
- then use e.g. intercept and slope to characterise groups of countries

```{r echo=FALSE}
library(dplyr)
library(tidyr)
library(purrr)

gapminder2 <- gapminder %>% mutate(year = year-1950)
by_country <- gapminder2 %>% 
  group_by(continent, country) %>% 
  nest()

by_country <- by_country %>% mutate(
    model = purrr::map(data, ~ lm(lifeExp ~ year, data = .))
)

by_country <- by_country %>% unnest(model %>% purrr::map(broom::tidy))
country_coefs <- by_country %>% select(continent, country, term, estimate) %>% spread(term, estimate)

ggplot(data=country_coefs, aes(x=`(Intercept)`, y=year, colour=continent)) +
  geom_point() + 
  scale_colour_brewer(palette="Dark2") + 
  xlab("Average Life Expectancy in 1950") +
  ylab("Average Gain in Life Expectancy per Year")
```


---
# Life Expectancy in the U.S.

Start with a single country, then generalize over all countries: 

```{r}
usa <- gapminder %>% filter(country=="United States")
head(usa)
```

---
# Life Expectancy in the U.S. since 1950

```{r}
ggplot(data=usa, aes(x=year, y=lifeExp)) + 
  geom_point() +
  geom_smooth(method="lm")
```

---
# Life Expectancy in the U.S. since 1950

```{r}
usa.lm <- lm(lifeExp~year, data=usa)
usa.lm
```

Intercept is estimated life expectancy at 0 BC - let's use 1950 for the first value:

```{r}
gapminder <- gapminder %>% mutate(year = year-1950)
```

---
# Nesting data

We don't want to subset the data for every country ... 

`nest()` makes a data frame part of another data frame:

```{r}
by_country <- gapminder %>% group_by(continent, country) %>% nest()
head(by_country)
```

Each element of the `data` variable in `by_country` is a dataset.

We will fit a linear model to each one of the elements of the `by_country` data.


---
# Fitting multiple models

Now we are using the `map` function in the package `purrr`.

`map` allows us to apply a function to each element of a list.

```{r}
by_country$model <- by_country$data %>% purrr::map(~lm(lifeExp~year, data=.))
head(by_country)
```

---
# `broom` again

using `broom::tidy` we extract coefficients for each model

```{r}
broom::tidy(by_country$model[[1]])
```

---
# Extract values for each coefficient

Extract all countries automatically (hello `map` again!)

```{r}
by_country$coefs <- by_country$model %>% purrr::map(broom::tidy)
by_country
```

---
# Get the coefficients into one data set

`unnest` pulls little data frames back into the whole frame

```{r}
by_country_coefs <- by_country %>% unnest(coefs)
by_country_coefs

```

---
# ... and finally, a visualization:

```{r}
coefs <- by_country_coefs %>% 
  select(country, continent, term, estimate) %>% 
  spread(term, estimate)
ggplot(data=coefs, aes(x=`(Intercept)`, y=year, colour=continent)) +
  geom_point()
```

---
class: inverse middle 
# Your turn

![](lorikeets.png)

- use `broom::glance` to get model fits for each one of the linear models of life expectancy by year
- which countries have the best five fits? which countries have the worst fits?
- are there meaningful patterns?

```{r, echo=FALSE, eval=FALSE}
by_country <- by_country %>% unnest(model %>% purrr::map(broom::glance))
by_country$country <- reorder(by_country$country, by_country$r.squared)
ggplot(data=by_country, aes(x=country, y=r.squared, colour=continent)) +
  geom_point() +
  theme(axis.text.x=element_text(angle=-90, hjust=0)) +
  scale_colour_brewer(palette="Dark2")
```


---
class: inverse middle 
# Your turn

![](lorikeets.png)

- extract residuals for each of the models and store it in a dataset together with country and continent information
- plot residuals across the years and fit a smooth. What does the pattern mean?

```{r eval=FALSE, echo=FALSE}
by_country$resids <- by_country$model %>% purrr::map(.f=residuals)
by_country_fits <- unnest(by_country, data, resids)
ggplot(data=by_country_fits, aes(x = lifeExp, y=resids)) + geom_point() +  geom_smooth()
# overall, a linear fit is not great: at low and high life expectancy, linear fit is too fast
```      

---
# broom for bio data

`biobroom` package (Bioconductor; Bass, Nelson, Robinson, Storey, 2016) has the same functions as `broom`, i.e. `glance`, `tidy`, and `augment`.

These functions provide information depending on the input type

```{r}
library(biobroom)
data(hammer)

counts <- Biobase::exprs(hammer)
head(counts)
```

---
# The Hammer study

published as part of the `biobroom` package, part of the [ReCount project](http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-449)
```{r}
# more information about the study:
Biobase::phenoData(hammer)@data
```

---
# biobroom and edgeR

identify differentially expressed genes following the protocol by Robinson, McCarthy and Smyth 2009

```{r}
library(edgeR)
y <- DGEList(counts = counts, group=Biobase::phenoData(hammer)@data$protocol)
y <- calcNormFactors(y)
y <- estimateCommonDisp(y)
y <- estimateTagwiseDisp(y)
et <- exactTest(y)

glance(et, alpha = 0.05) # glance on DGEExact
```

---
# biobroom and edgeR

```{r}
tet <- broom::tidy(et)
tet$significant <- tet$p.value < 0.05
ggplot(data=tet, aes(x=logCPM, y=estimate, colour=significant)) +
  geom_point(alpha=.1) + facet_wrap(~significant, labeller="label_both")
```

`augment(y)` only works for DGEList objects


---
# biobroom and limma

- ... works similarly to biobroom and edgeR ...
- we will pick up on this in the multivariate session after lunch

---
# References

- Ian Lyttle's notes on using [purrr ](http://ijlyttle.github.io/isugg_purrr/presentation.html#(1)
- RStudio's tutorial by Charlotte Wickham on using [purrr](https://www.rstudio.com/resources/videos/happy-r-users-purrr-tutorial/)
- [R blogger's article on purrr](http://www.r-bloggers.com/purrr-0-1-0/)
- Data transformation [cheatsheet](https://github.com/rstudio/cheatsheets/raw/master/source/pdfs/data-transformation-cheatsheet.pdf)
- [Wickham (2007) Reshaping data](https://www.jstatsoft.org/article/view/v021i12/v21i12.pdf)
- [broom vignettes, David Robinson](https://cran.r-project.org/web/packages/broom/vignettes/broom.html)
- [Wickham (2011) Split-Apply-Combine](https://www.jstatsoft.org/article/view/v040i01)

---
# Share and share alike

This work is licensed under the Creative Commons Attribution-Noncommercial 3.0 United States License. To view a copy of this license, visit http://creativecommons.org/licenses/by-nc/3.0/us/ or send a letter to Creative Commons, 171 Second Street, Suite 300, San Francisco, California, 94105, USA.
